lib/cfg.js

```diff
@@ -0,0 +1,368 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+
+var _ = require('underscore-node');
+
+function ControlFlowGraph(rules){
+    this.rules = rules;
+    
+}
+
+function seperateCodeFromData(asmArr) {
+    
+    var isCode = function(asmArr, index){
+        return (((asmArr[index].source !== null) &&
+                (asmArr[index+1].source !== null)) ||
+                ((asmArr[index].source === null) &&
+                 (asmArr[index+1].source !== null)) ||
+                 ((asmArr[index].source !== null) &&
+                 (asmArr[index+1].source === null)));
+    };
+
+    var code = [];
+    var first = 0;
+    var last = asmArr.length;
+    
+    if(first == last) return code;
+
+    
+    while(first != last-1){
+        if(isCode(asmArr, first))
+            code.push(JSON.parse(JSON.stringify(asmArr[first])));
+        
+        ++first;
+    }
+    if(asmArr[first].source){
+        code.push(JSON.parse(JSON.stringify(asmArr[first])));
+    }
+
+    return code;
+}
+
+    var gccX86 = {
+                filterData:  seperateCodeFromData,
+                isFunctionEnd:function(x)  { return ((x[0] != ' ') && (x[0] != '.') &&
+                                                             (x.indexOf(':') != -1)) ;},
+
+                isBasicBlockEnd: function(x) { return x[0] == ".";} ,
+
+                getInstructionType:function(inst) {
+                                            inst = inst.trim();
+                                            if (inst === "") return -1;
+                                            else if(inst.includes("jmp")) return 0;
+                                            else if(inst[0] === 'j') return 1;
+                                            else if(!inst.includes("ret")) return 2;
+                                            else return 3;
+                                          },
+
+                extractNodeName:function(inst) {
+                                     var name = inst.match(/.L\d+/);
+                                     return name + ":";
+                                  },
+
+                isJmpInstruction: function(x) { var trimed = x.trim(); return ( trimed[0] == 'j');}
+             };
+
+
+
+
+ControlFlowGraph.prototype.splitToFunctions = function (asmArr, isEnd) {
+    if(asmArr.length === 0)   return [];
+    var result = [];
+    var first = 0;
+    var last = asmArr.length;
+    var fnRange = {start:first, end:null};
+    ++first;
+    while(first != last) {
+        if(isEnd(asmArr[first].text)) {
+            fnRange.end = first;
+            result.push(JSON.parse(JSON.stringify(fnRange)));
```

Comment:
It seems super odd to JSON parse and JSON stringify this. Given your comment below I think you're trying to make a deep copy here, right? See `_.clone()` for a more intention-revealing way to do this.

Reply:
done